File: lib/store/round-store.ts
"use client";

import { create } from "zustand";
import { calcRoundTotal } from "@/lib/domain/analytics";
import { DEFAULT_UI_PREFERENCES } from "@/lib/domain/constants";
import { createRoundRecord, mergeConfig } from "@/lib/domain/rounds";
import { isRoundReadyToComplete } from "@/lib/domain/progress";
import type { GolfCourse, HoleScore, RoundConfig, RoundRecord, RoundType, UiPreferences } from "@/lib/domain/types";
import { getInProgressRound, listCourses, listRoundRecords, saveCourse, saveRoundRecord, db, getSettings, saveSettings } from "@/lib/db/app-db";

type RoundStore = {
  currentRound: RoundRecord | null;
  rounds: RoundRecord[];
  courses: GolfCourse[];
  roundDefaults: RoundConfig;
  uiPreferences: UiPreferences;
  hydrated: boolean;
  boot: () => Promise<void>;
  startRound: (input: {
    playedAt: string;
    courseId?: string | null;
    courseName: string;
    roundType: RoundType;
    holePars?: number[];
  }) => Promise<void>;
  updateHole: (holeNo: number, patch: Partial<HoleScore>) => Promise<void>;
  goToHole: (holeNo: number) => Promise<void>;
  completeRound: () => Promise<boolean>;
  saveCourseMaster: (course: GolfCourse) => Promise<void>;
  saveDefaults: (config: RoundConfig) => Promise<void>;
  saveUiPreferences: (preferences: Partial<UiPreferences>) => Promise<void>;
};

function touchRound(record: RoundRecord): RoundRecord {
  const totalScore = calcRoundTotal(record.holes);

  return {
    round: {
      ...record.round,
      totalScore,
      lastInputAt: new Date().toISOString(),
    },
    holes: record.holes,
  };
}

export const useRoundStore = create<RoundStore>((set, get) => ({
  currentRound: null,
  rounds: [],
  courses: [],
  roundDefaults: mergeConfig(),
  uiPreferences: DEFAULT_UI_PREFERENCES,
  hydrated: false,
  async boot() {
    const [records, currentRound, settings, courses] = await Promise.all([
      listRoundRecords(),
      getInProgressRound(),
      getSettings(),
      listCourses(),
    ]);

    let currentRecord: RoundRecord | null = null;
    if (currentRound) {
      const currentHoles = await db.holes
        .where("roundId")
        .equals(currentRound.id)
        .sortBy("holeNo");
      currentRecord = {
        round: currentRound,
        holes: currentHoles,
      };
    }

    set({
      rounds: records,
      courses,
      currentRound: currentRecord,
      roundDefaults: mergeConfig(settings?.roundDefaults),
      uiPreferences: {
        ...DEFAULT_UI_PREFERENCES,
        ...(settings?.uiPreferences ?? {}),
      },
      hydrated: true,
    });
  },
  async startRound(input) {
    const configSnapshot = get().roundDefaults;
    const record = createRoundRecord({
      ...input,
      configSnapshot,
      courseName: input.courseName,
    });

    await saveRoundRecord(record);
    set((state) => ({
      currentRound: record,
      rounds: [record, ...state.rounds.filter((item) => item.round.id !== record.round.id)],
    }));
  },
  async updateHole(holeNo, patch) {
    const current = get().currentRound;
    if (!current) return;

    const holes = current.holes.map((hole) =>
      hole.holeNo === holeNo
        ? {
            ...hole,
            ...patch,
            updatedAt: new Date().toISOString(),
          }
        : hole,
    );

    const record = touchRound({
      round: current.round,
      holes,
    });

    set((state) => ({
      currentRound: record,
      rounds: [record, ...state.rounds.filter((item) => item.round.id !== record.round.id)],
    }));

    try {
      await saveRoundRecord(record);
    } catch (error) {
      console.error("Failed to save hole update", error);
    }
  },
  async goToHole(holeNo) {
    const current = get().currentRound;
    if (!current) return;

    const nextHoleNo = Math.min(Math.max(1, holeNo), current.round.holesCount);
    const record: RoundRecord = {
      round: {
        ...current.round,
        currentHoleNo: nextHoleNo,
        lastInputAt: new Date().toISOString(),
      },
      holes: current.holes,
    };

    set((state) => ({
      currentRound: record,
      rounds: [record, ...state.rounds.filter((item) => item.round.id !== record.round.id)],
    }));

    try {
      await saveRoundRecord(record);
    } catch (error) {
      console.error("Failed to save hole move", error);
    }
  },
  async completeRound() {
    const current = get().currentRound;
    if (!current) return false;
    if (!isRoundReadyToComplete(current)) return false;

    const completedAt = new Date().toISOString();
    const record: RoundRecord = {
      round: {
        ...current.round,
        status: "COMPLETED",
        finishedAt: completedAt,
        totalScore: calcRoundTotal(current.holes),
        lastInputAt: completedAt,
      },
      holes: current.holes,
    };

    await saveRoundRecord(record);

    set((state) => ({
      currentRound: null,
      rounds: [record, ...state.rounds.filter((item) => item.round.id !== record.round.id)],
    }));

    return true;
  },
  async saveCourseMaster(course) {
    await saveCourse(course);
    const courses = await listCourses();
    set({ courses });
  },
  async saveDefaults(config) {
    const merged = mergeConfig(config);
    await saveSettings({ roundDefaults: merged });
    set({ roundDefaults: merged });
  },
  async saveUiPreferences(preferences) {
    const merged = {
      ...get().uiPreferences,
      ...preferences,
    };
    await saveSettings({ uiPreferences: merged });
    set({ uiPreferences: merged });
  },
}));

File: flutter_app/lib/state/round_store.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';

import '../domain/analytics.dart';
import '../domain/constants.dart';
import '../domain/courses.dart';
import '../domain/models.dart';
import '../domain/progress.dart';
import '../domain/rounds.dart';

class RoundStore extends ChangeNotifier {
  static const _roundsKey = 'round_records_v1';
  static const _coursesKey = 'courses_v1';
  static const _settingsKey = 'settings_v1';

  RoundRecord? currentRound;
  List<RoundRecord> rounds = [];
  List<GolfCourse> courses = [];
  RoundConfig roundDefaults = mergeConfig();
  UiPreferences uiPreferences = defaultUiPreferences;
  bool hydrated = false;

  Future<void> boot() async {
    final prefs = await SharedPreferences.getInstance();

    final roundsRaw = prefs.getString(_roundsKey);
    if (roundsRaw != null && roundsRaw.isNotEmpty) {
      final decoded = (jsonDecode(roundsRaw) as List<dynamic>)
          .map((entry) => RoundRecord.fromJson(entry as Map<String, dynamic>))
          .toList();
      rounds = decoded;
    } else {
      rounds = [];
    }

    final coursesRaw = prefs.getString(_coursesKey);
    if (coursesRaw != null && coursesRaw.isNotEmpty) {
      courses = (jsonDecode(coursesRaw) as List<dynamic>)
          .map((entry) => GolfCourse.fromJson(entry as Map<String, dynamic>))
          .toList();
    } else {
      courses = List<GolfCourse>.from(courseCatalog);
    }

    final settingsRaw = prefs.getString(_settingsKey);
    if (settingsRaw != null && settingsRaw.isNotEmpty) {
      final settings = jsonDecode(settingsRaw) as Map<String, dynamic>;
      roundDefaults = RoundConfig.fromJson(
        settings['roundDefaults'] as Map<String, dynamic>? ??
            defaultRoundConfig.toJson(),
      );
      uiPreferences = UiPreferences.fromJson(
        settings['uiPreferences'] as Map<String, dynamic>? ??
            defaultUiPreferences.toJson(),
      );
    } else {
      roundDefaults = mergeConfig();
      uiPreferences = defaultUiPreferences;
    }

    currentRound = _findInProgressRound(rounds);
    hydrated = true;
    notifyListeners();
  }

  Future<void> startRound({
    required String playedAt,
    String? courseId,
    required String courseName,
    required RoundType roundType,
    List<int>? holePars,
  }) async {
    final record = createRoundRecord(
      CreateRoundInput(
        playedAt: playedAt,
        courseId: courseId,
        courseName: courseName,
        roundType: roundType,
        configSnapshot: roundDefaults,
        holePars: holePars,
      ),
    );

    currentRound = record;
    rounds = [
      record,
      ...rounds.where((entry) => entry.round.id != record.round.id)
    ];

    await _persist();
    notifyListeners();
  }

  Future<void> updateHole(
    int holeNo, {
    int? strokes,
    int? putts,
    TeeShotDirection? teeShotDirection,
    bool clearDirection = false,
    bool? puttsUnknown,
    int? obCount,
    int? otherPenaltyCount,
    bool? fwKeep,
    bool clearFwKeep = false,
    bool? gir,
    bool clearGir = false,
    bool? bunkerIn,
    bool clearBunkerIn = false,
    int? bunkerCount,
  }) async {
    final current = currentRound;
    if (current == null) return;

    final now = DateTime.now().toIso8601String();

    final holes = current.holes.map((hole) {
      if (hole.holeNo != holeNo) return hole;

      final nextObCount = obCount ?? hole.obCount;
      final nextOtherCount = otherPenaltyCount ?? hole.otherPenaltyCount;

      return hole.copyWith(
        strokes: strokes,
        putts: putts,
        teeShotDirection: teeShotDirection,
        clearDirection: clearDirection,
        puttsUnknown: puttsUnknown,
        obCount: nextObCount,
        otherPenaltyCount: nextOtherCount,
        penaltyTotal: nextObCount + nextOtherCount,
        fwKeep: fwKeep,
        clearFwKeep: clearFwKeep,
        gir: gir,
        clearGir: clearGir,
        bunkerIn: bunkerIn,
        clearBunkerIn: clearBunkerIn,
        bunkerCount: bunkerCount,
        updatedAt: now,
      );
    }).toList();

    final record = _touchRound(RoundRecord(round: current.round, holes: holes));
    currentRound = record;
    rounds = [
      record,
      ...rounds.where((entry) => entry.round.id != record.round.id)
    ];

    await _persist();
    notifyListeners();
  }

  Future<void> goToHole(int holeNo) async {
    final current = currentRound;
    if (current == null) return;

    final nextHoleNo = holeNo.clamp(1, current.round.holesCount);
    final nextRound = current.round.copyWith(
      currentHoleNo: nextHoleNo,
      lastInputAt: DateTime.now().toIso8601String(),
    );

    final record = RoundRecord(round: nextRound, holes: current.holes);
    currentRound = record;
    rounds = [
      record,
      ...rounds.where((entry) => entry.round.id != record.round.id)
    ];

    await _persist();
    notifyListeners();
  }

  Future<bool> completeRound() async {
    final current = currentRound;
    if (current == null) return false;
    if (!isRoundReadyToComplete(current)) return false;

    final completedAt = DateTime.now().toIso8601String();

    final nextRound = current.round.copyWith(
      status: RoundStatus.completed,
      finishedAt: completedAt,
      totalScore: calcRoundTotal(current.holes),
      lastInputAt: completedAt,
    );

    final completedRecord = RoundRecord(round: nextRound, holes: current.holes);

    currentRound = null;
    rounds = [
      completedRecord,
      ...rounds.where((entry) => entry.round.id != completedRecord.round.id),
    ];

    await _persist();
    notifyListeners();
    return true;
  }

  Future<void> saveCourseMaster(GolfCourse course) async {
    final nextCourses = [
      course,
      ...courses.where((entry) => entry.id != course.id),
    ]..sort((a, b) => a.name.compareTo(b.name));

    courses = nextCourses;
    await _persist();
    notifyListeners();
  }

  Future<void> saveDefaults(RoundConfig config) async {
    roundDefaults = mergeConfig(config);
    await _persist();
    notifyListeners();
  }

  Future<void> saveUiPreferences(UiPreferences preferences) async {
    uiPreferences = preferences;
    await _persist();
    notifyListeners();
  }

  RoundRecord? _findInProgressRound(List<RoundRecord> records) {
    for (final record in records) {
      if (record.round.status == RoundStatus.inProgress) {
        return record;
      }
    }
    return null;
  }

  RoundRecord _touchRound(RoundRecord record) {
    final total = calcRoundTotal(record.holes);
    final touchedRound = record.round.copyWith(
      totalScore: total,
      lastInputAt: DateTime.now().toIso8601String(),
    );
    return RoundRecord(round: touchedRound, holes: record.holes);
  }

  Future<void> _persist() async {
    final prefs = await SharedPreferences.getInstance();

    await prefs.setString(
      _roundsKey,
      jsonEncode(rounds.map((entry) => entry.toJson()).toList()),
    );

    await prefs.setString(
      _coursesKey,
      jsonEncode(courses.map((entry) => entry.toJson()).toList()),
    );

    await prefs.setString(
      _settingsKey,
      jsonEncode({
        'roundDefaults': roundDefaults.toJson(),
        'uiPreferences': uiPreferences.toJson(),
      }),
    );
  }
}
